File: verifier/src/main/kotlin/tech/thessemaj/deviceintelligence/verifier/TokenVerifier.kt
package tech.thessemaj.deviceintelligence.verifier

import java.security.Signature
import java.security.cert.X509Certificate

/**
 * The backend entry point: turn a token + the nonce the server issued into a
 * trust [Decision]. Kotlin port of tools/server/verify_token.py.
 *
 * ```
 * val result = TokenVerifier().verify(tokenHex, issuedNonce)
 * when (result.decision) {
 *     Decision.TRUSTWORTHY -> allow()
 *     Decision.COMPROMISED -> stepUp(result.blockingSignals)
 *     Decision.REJECT      -> deny()   // not a genuine, fresh binding
 * }
 * ```
 *
 * The decision is LAYERED and mirrors the Python exactly:
 *  - authenticity — binding present · nonce fresh · challenge == nonce ·
 *    chain → pinned Google root · ECDSA signature valid. Fail any ⇒ REJECT.
 *  - device integrity — the TEE's own attestation: security level ≥ TEE,
 *    verifiedBootState == Verified, deviceLocked.
 *  - policy over signals — each resolved code run through [Policy].
 *
 * `decision = authentic AND deviceIntegrityOk AND no blocking signal`.
 */
class TokenVerifier(
    private val registry: SignalRegistry = SignalRegistry.bundled,
    private val policy: Policy = Policy(),
    pinnedRoots: List<X509Certificate> = PinnedRoots.default,
) {
    private val chainVerifier = ChainVerifier(pinnedRoots)

    fun verify(tokenHex: String, issuedNonce: String): VerificationResult {
        val checks = ArrayList<Check>()
        fun auth(name: String, ok: Boolean, detail: String = "") = ok.also { checks.add(Check(name, ok, detail, CheckKind.AUTH)) }
        fun integ(name: String, ok: Boolean, detail: String = "") = ok.also { checks.add(Check(name, ok, detail, CheckKind.INTEGRITY)) }

        val text = Keystream.decryptHex(tokenHex)
        val sepIdx = text.indexOf(TokenDecoder.BINDING_SEP)
        val signed = if (sepIdx >= 0) text.substring(0, sepIdx) else text
        val binding = if (sepIdx >= 0) text.substring(sepIdx + TokenDecoder.BINDING_SEP.length) else ""

        // Parse the document up front so we can always surface the (untrusted) signals.
        val doc: Map<String, Any?> = runCatching { Json.parseObject(signed) }.getOrDefault(emptyMap())

        if (!auth("binding present", sepIdx >= 0, if (sepIdx < 0) "unbound/legacy token" else "")) {
            return result(checks, doc)
        }
        if (doc.isEmpty()) {
            auth("signed content is JSON", false, "unparseable signed_content")
            return result(checks, doc)
        }

        val tokenNonce = doc["nonce"] as? String ?: ""
        auth("nonce matches issued", tokenNonce == issuedNonce, "token=${tokenNonce.take(16)}… issued=${issuedNonce.take(16)}…")

        // Parse binding: SIG<FS>hex and one-or-more CERT<FS>hex lines.
        var sigHex = ""
        val certsHex = ArrayList<String>()
        for (line in binding.split("\n")) {
            when {
                line.startsWith("SIG$FS") -> sigHex = line.substring(4)
                line.startsWith("CERT$FS") -> certsHex.add(line.substring(5))
            }
        }
        if (!auth("chain + signature present", sigHex.isNotEmpty() && certsHex.isNotEmpty())) {
            return result(checks, doc)
        }

        val chain = runCatching { chainVerifier.parseChain(certsHex) }.getOrNull()
        if (chain == null || chain.isEmpty()) {
            auth("chain parses", false, "could not parse cert chain")
            return result(checks, doc)
        }
        val leaf = chain.first()

        runCatching { chainVerifier.verifyToPinnedRoot(chain) }
            .onSuccess { auth("chain -> pinned Google root", true, it.subjectX500Principal.name) }
            .onFailure { auth("chain -> pinned Google root", false, it.message ?: "chain error") }

        runCatching { Attestation.challenge(leaf) }
            .onSuccess { chal -> auth("attestation challenge == nonce", Hex.encode(chal) == issuedNonce.lowercase(), "challenge=${Hex.encode(chal).take(16)}…") }
            .onFailure { auth("attestation challenge == nonce", false, it.message ?: "no challenge") }

        runCatching {
            val sig = Signature.getInstance("SHA256withECDSA")
            sig.initVerify(leaf.publicKey)
            sig.update(signed.toByteArray(Charsets.UTF_8))
            sig.verify(Hex.decode(sigHex))
        }.onSuccess { ok -> auth("signature over verdict", ok, if (ok) "" else "ECDSA verify failed") }
            .onFailure { auth("signature over verdict", false, it.message ?: "signature error") }

        // Device-integrity layer — the TEE's own attestation fields.
        runCatching { Attestation.fields(leaf) }
            .onSuccess { f ->
                integ("hardware security level >= TEE", f.securityLevel == 1 || f.securityLevel == 2, f.securityLevelName)
                integ("verified boot state = Verified", f.verifiedBootState == 0, f.bootStateName)
                integ("device locked", f.deviceLocked == true, f.deviceLocked.toString())
            }
            .onFailure { integ("attestation device-integrity fields", false, it.message ?: "parse error") }

        return result(checks, doc)
    }

    private fun result(checks: List<Check>, doc: Map<String, Any?>): VerificationResult {
        val authentic = checks.filter { it.kind == CheckKind.AUTH }.all { it.ok }
        val deviceOk = checks.filter { it.kind == CheckKind.INTEGRITY }.all { it.ok }
        val signals = Signals.resolve(doc, registry, policy)
        val blocking = signals.any { it.blocking }
        val decision = when {
            !authentic -> Decision.REJECT
            deviceOk && !blocking -> Decision.TRUSTWORTHY
            else -> Decision.COMPROMISED
        }
        return VerificationResult(
            decision = decision,
            authentic = authentic,
            deviceIntegrityOk = deviceOk,
            checks = checks,
            schemaVersion = (doc["schemaVersion"] as? Long)?.toInt(),
            point = doc["point"] as? String,
            ts = doc["ts"] as? Long,
            nonce = doc["nonce"] as? String,
            device = Signals.device(doc),
            signals = signals,
        )
    }

    fun verifyChallenge(tokenHex: String, issuedChallenge: String, sessionSigner: SessionSigner): VerificationResult {
        val checks = ArrayList<Check>()
        fun auth(n: String, ok: Boolean, d: String = "") = ok.also { checks.add(Check(n, ok, d, CheckKind.AUTH)) }
        fun integ(n: String, ok: Boolean, d: String = "") = ok.also { checks.add(Check(n, ok, d, CheckKind.INTEGRITY)) }

        val text = Keystream.decryptHex(tokenHex)
        val sepIdx = text.indexOf(TokenDecoder.BINDING_SEP)
        val signed = if (sepIdx >= 0) text.substring(0, sepIdx) else text
        val binding = if (sepIdx >= 0) text.substring(sepIdx + TokenDecoder.BINDING_SEP.length) else ""
        val doc: Map<String, Any?> = runCatching { Json.parseObject(signed) }.getOrDefault(emptyMap())

        val session = (doc["sessionId"] as? String)?.let { sessionSigner.open(it) }
        if (!auth("session valid", session != null, if (session == null) "sessionId forged/expired -> re-enroll" else "")) {
            return challengeResult(checks, doc, null)
        }
        val tokenChallenge = doc["challenge"] as? String ?: ""
        auth("challenge matches issued", tokenChallenge == issuedChallenge, "token=${tokenChallenge.take(12)}… issued=${issuedChallenge.take(12)}…")

        var sigHex = ""
        for (line in binding.split("\n")) if (line.startsWith("SIG$FS")) sigHex = line.substring(4)
        if (!auth("signature present", sigHex.isNotEmpty())) return challengeResult(checks, doc, session)

        runCatching {
            val spki = java.security.spec.X509EncodedKeySpec(Hex.decode(session!!.pinnedKeySpkiHex))
            val pub = java.security.KeyFactory.getInstance("EC").generatePublic(spki)
            Signature.getInstance("SHA256withECDSA").run {
                initVerify(pub); update(signed.toByteArray(Charsets.UTF_8)); verify(Hex.decode(sigHex))
            }
        }.onSuccess { ok -> auth("signature by pinned key", ok, if (ok) "" else "ECDSA verify failed") }
            .onFailure { auth("signature by pinned key", false, it.message ?: "sig error") }

        // All attestation validations run HERE, on the facts carried (HMAC-signed) from
        // enroll — enroll itself never rejects on them. Proven forgeries are AUTH failures
        // (-> REJECT: the attestation isn't genuine); honest-but-compromised device-integrity
        // facts are INTEGRITY failures (-> COMPROMISED). This is where the backend decides.
        session?.let {
            // Forgeries / active spoofers -> REJECT.
            auth("attestation chain trusted", it.chainTrusted, if (it.chainTrusted) "" else "chain does not reach a pinned Google root")
            auth("no revoked keybox", !it.keyboxRevoked)
            auth("no cross-level keybox reuse", !it.crossLevelReuse)
            // strongboxChainMissing is NOT an auth REJECT: a genuine StrongBox device can hit a
            // transient StrongBox failure. It surfaces as INTEL_0045 (HIGH, observe-only; block via
            // policy if you want it hard). INTEL_0039 (capability interlock) is retired.
            auth("device-property attestation matches self-report", !it.devicePropMismatch)
            auth("boot-state self-report matches hardware attestation", !it.bootStateSpoofer,
                 if (it.bootStateSpoofer) "self-report claims clean/locked boot but attestation says otherwise — prop spoofer" else "")
            // Honest device-integrity facts -> COMPROMISED.
            // Real gate: SOFTWARE (which also covers a missing/unparseable securityLevel)
            // is not hardware-backed and fails. This was previously written as
            // `TEE || STRONGBOX` against a two-valued enum, i.e. always true.
            integ("hardware security level >= TEE", it.assurance != Assurance.SOFTWARE, it.assurance.name)
            if (policy.requireStrongBox) integ("StrongBox required by policy", it.assurance == Assurance.STRONGBOX, it.assurance.name)
            integ("verified boot state = Verified", it.bootState == "Verified", it.bootState)
            integ("device locked", it.deviceLocked, it.deviceLocked.toString())
        }
        return challengeResult(checks, doc, session)
    }

    private fun challengeResult(checks: List<Check>, doc: Map<String, Any?>, session: Session?): VerificationResult {
        val authentic = checks.filter { it.kind == CheckKind.AUTH }.all { it.ok }
        val deviceOk = checks.filter { it.kind == CheckKind.INTEGRITY }.all { it.ok }
        val signals = Signals.resolve(doc, registry, policy) + attestationSignals(session)
        val decision = when {
            !authentic -> Decision.REJECT
            deviceOk && signals.none { it.blocking } -> Decision.TRUSTWORTHY
            else -> Decision.COMPROMISED
        }
        return VerificationResult(decision, authentic, deviceOk, checks,
            (doc["schemaVersion"] as? Long)?.toInt(), doc["name"] as? String,
            doc["ts"] as? Long, doc["challenge"] as? String, Signals.device(doc), signals)
    }

    /**
     * Surface the carried attestation forgeries as first-class registry SIGs, so a
     * boot-state spoof appears as INTEL_0055 in the
     * verdict's signal list (the shared taxonomy), alongside the auth() gate that
     * already REJECTs it. The gate is the authority; the SIG is the named telemetry.
     */
    private fun attestationSignals(session: Session?): List<ResolvedSignal> {
        if (session == null) return emptyList()
        val out = ArrayList<ResolvedSignal>()
        fun add(id: String, detail: String) = registry[id]?.let { m ->
            out.add(ResolvedSignal(m.id, m.detector, m.kind, m.title, m.severity, detail,
                policy.isBlocking(m.id, m.severity)))
        }
        if (session.bootStateSpoofer) add("INTEL_0055", "self-report=green/locked but hardware attestation disagrees")
        if (session.crossLevelReuse) add("INTEL_0016", "same attestation batch key across StrongBox and TEE — leaked keybox")
        if (session.strongboxChainMissing) add("INTEL_0045", "StrongBox hardware indicated but no StrongBox attestation chain produced (fail-closed)")
        if (session.softwareAttested) add("INTEL_0056", "attestation reports securityLevel=Software — no hardware root of trust")
        return out
    }

    private companion object {
        const val FS = "\u001F"
    }
}

File: verifiers/python/src/deviceintelligence_verifier/token_verifier.py
"""The v1-era verify flow (TokenVerifier.kt port): authenticity + TEE facts."""
import hashlib, json as _json
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import ec
from .keystream import decrypt_hex
from .models import Check, CheckKind, Decision, VerificationResult
from .chain_verifier import parse_chain as _parse_chain
from .pinned_roots import default as _pinned_default
from .attestation import challenge as _att_challenge, fields as _att_fields
from .signals import resolve as _resolve, device as _device
from .codec import decode as _codec_decode

BINDING_SEP = "\n--BINDING\n"
FS = "\x1F"


class TokenVerifier:
    def __init__(self, registry=None, policy=None, pinned_roots=None):
        from .registry import SignalRegistry
        from .policy import Policy
        self.registry = registry or SignalRegistry.bundled()
        self.policy = policy or Policy()
        self.pinned_roots = pinned_roots or _pinned_default()

    def _parse_chain(self, certs_hex):
        return _parse_chain(certs_hex)

    def verify(self, token_hex: str, issued_nonce: str) -> VerificationResult:
        checks = []
        def auth(name, ok, detail=""):
            checks.append(Check(name, ok, detail, CheckKind.AUTH)); return ok
        def integ(name, ok, detail=""):
            checks.append(Check(name, ok, detail, CheckKind.INTEGRITY)); return ok

        text = decrypt_hex(token_hex)
        sep = text.find(BINDING_SEP)
        signed = text[:sep] if sep >= 0 else text
        binding = text[sep + len(BINDING_SEP):] if sep >= 0 else ""
        doc = _run(lambda: _json.loads(signed)).get("value") or {}

        if not auth("binding present", sep >= 0, "" if sep >= 0 else "unbound/legacy token"):
            return self._result(checks, doc)
        if not doc:
            auth("signed content is JSON", False, "unparseable signed_content")
            return self._result(checks, doc)

        token_nonce = doc.get("nonce") or ""
        auth("nonce matches issued", token_nonce == issued_nonce)

        sig_hex, certs_hex = "", []
        for line in binding.split("\n"):
            if line.startswith("SIG" + FS): sig_hex = line[4:]
            elif line.startswith("CERT" + FS): certs_hex.append(line[5:])
        if not auth("chain + signature present", bool(sig_hex) and bool(certs_hex)):
            return self._result(checks, doc)

        chain = _run(lambda: _parse_chain(certs_hex)).get("value")
        if not chain:
            auth("chain parses", False, "could not parse cert chain")
            return self._result(checks, doc)
        leaf = chain[0]

        root_err = _run(lambda: self._chain.verify_to_pinned_root(chain) if hasattr(self, "_cv") else
                        _verify_to_pinned(chain, self.pinned_roots))
        auth("chain -> pinned Google root", root_err.get("ok"),
             root_err.get("detail", "") if not root_err.get("ok") else _subject(root_err.get("value")))

        chal = _run(lambda: _att_challenge(leaf))
        auth("attestation challenge == nonce",
             chal.get("ok") and chal["value"].hex() == issued_nonce.lower())

        sig_ok = _run(lambda: leaf.public_key().verify(
            bytes.fromhex(sig_hex), signed.encode("utf-8"),
            ec.ECDSA(hashes.SHA256()))).get("ok", False)
        auth("signature over verdict", sig_ok, "" if sig_ok else "ECDSA verify failed")

        f = _run(lambda: _att_fields(leaf)).get("value")
        integ("hardware security level >= TEE", f is not None and f.security_level in (1, 2),
              f.security_level_name if f else "parse error")
        integ("verified boot state = Verified", f is not None and f.verified_boot_state == 0,
              f.boot_state_name if f else "parse error")
        integ("device locked", f is not None and f.device_locked is True,
              str(f.device_locked) if f else "parse error")

        return self._result(checks, doc)

    def _result(self, checks, doc):
        authentic = all(c.ok for c in checks if c.kind == CheckKind.AUTH)
        device_ok = all(c.ok for c in checks if c.kind == CheckKind.INTEGRITY)
        signals = _resolve(doc, self.registry, self.policy)
        blocking = any(s.blocking for s in signals)
        decision = (Decision.REJECT if not authentic else
                    Decision.COMPROMISED if (not device_ok or blocking) else
                    Decision.TRUSTWORTHY)
        return VerificationResult(decision, authentic, device_ok, checks,
                                  doc.get("schemaVersion"), doc.get("point"), doc.get("ts"),
                                  doc.get("nonce"), _device(doc), signals)


def _run(fn):
    try:
        return {"ok": True, "value": fn()}
    except Exception as e:
        return {"ok": False, "error": str(e)}


def _verify_to_pinned(chain, pinned_roots):
    from .chain_verifier import verify_to_pinned_root
    return verify_to_pinned_root(chain, pinned_roots)


def _subject(cert):
    return cert.subject.rfc4514_string()


def _att_challenge(leaf):
    from .attestation import challenge
    return challenge(leaf)


def _att_fields(leaf):
    from .attestation import fields
    return fields(leaf)
